subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/DefaultVersionedComponentChooser.java

```diff
@@ -109,22 +109,28 @@ public void selectNewestMatchingComponent(Collection<? extends ModuleComponentRe
                 continue;
             }
 
-            RejectedByAttributesVersion maybeRejectByAttributes = tryRejectByAttributes(candidateId, metadataProvider, consumerAttributes);
-            if (maybeRejectByAttributes != null) {
-                result.doesNotMatchConsumerAttributes(maybeRejectByAttributes);
-            } else if (isRejectedBySelector(candidateId, rejectedVersionSelector)) {
+            // Do this check first, it cannot require the metadata
```

Comment:
Can we move the versionMatches call above down? Since it looks like it sometimes uses the metadata as well

Reply:
Not too keen on doing this change. It could cause users to see a "rejected" version before seeing it did not match.

But this will indeed happen if you use a prefix version like `1.+`. We should figure out why we think this selector needs the metadata.